### PR TITLE
Closure fixes

### PR DESCRIPTION
**Pull request by sgomes**
This PR updates the codebase to play better with the Closure compiler and JSDoc.
It also tweaks the JSCS config, updates the JSCS version and adds a custom linting rule, all in to enforce checks a bit better.

This PR shouldn't break BC.

@addyosmani @samthor PTAL!

### CODE REVIEW

**Comment by Garbee** [comment]
LGTM. No BC concerns either, looks like anything that *may* cause problems isn't supported usage in the first place (since mods are to internal properties.)

**Comment by addyosmani** [comment]
LGTM. Deferring to @samthor for a final LGTM before we move ahead.

File: gulpfile.js [hunk 1]
```diff
@@ -566,8 +572,11 @@ gulp.task('publish:code', function(cb) {
     cb);
 });
 
-// Function to publish staging or prod version from local tree,
-// or to promote staging to prod, per passed arg.
+/**
+ * Function to publish staging or prod version from local tree,
+ * or to promote staging to prod, per passed arg.
+ * @param  {string} pubScope the scope to publish to.
```
Comment (samthor) [inline gulpfile.js]: nit: single space between @param and {string}

File: src/layout/layout.js [hunk 2]
```diff
@@ -402,6 +408,15 @@
     }
   };
 
+  /**
+   * Constructor for an individual tab.
+   *
+   * @constructor
+   * @param {HTMLElement} tab The HTML element for the tab.
+   * @param {Array.<HTMLElement>} tabs Array with HTML elements for all tabs.
```
Comment (samthor) [inline src/layout/layout.js]: Not a big deal, but the '.' between Array and <HTMLElement> is now no longer required.

Comment (samthor) [inline src/layout/layout.js]: Also, if you're confident these will never be null, then you can prefix `Array` with !.

File: src/third_party/rAF.js [hunk 3]
```diff
@@ -25,14 +29,17 @@ for (var i = 0; i < vendors.length && !window.requestAnimationFrame; ++i) {
 
 if (/iP(ad|hone|od).*OS 6/.test(window.navigator.userAgent) || !window.requestAnimationFrame || !window.cancelAnimationFrame) {
   var lastTime = 0;
+  /**
+   * requestAnimationFrame polyfill.
+   * @param  {Function} callback the callback function.
```
Comment (samthor) [inline src/third_party/rAF.js]: should be !Function

File: utils/jscs-rules/closure-camel-case.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+ * Returns the option name to JSCS.
+ * @return {string} the option name.
+ */
+module.exports.prototype.getOptionName = function() {
+  return 'closureCamelCase';
+};
+
+/**
+ * Configure the rule parameters.
+ * @param  {!boolean} value the value for this rule.
```
Comment (samthor) [inline utils/jscs-rules/closure-camel-case.js]: ! is redundant, this is a native type

File: src/mdlComponentHandler.js [hunk 5]
```diff
@@ -87,20 +87,21 @@ window.componentHandler = (function() {
...
-   * @param {String=} optJsClass the programatic name of the element class we
+   * @param {string=} optJsClass the programatic name of the element class we
    * need to create a new instance of.
-   * @param {String=} optCssClass the name of the CSS class elements of this
+   * @param {string=} optCssClass the name of the CSS class elements of this
    * type will have.
    */
   function upgradeDomInternal(optJsClass, optCssClass) {
-    if (optJsClass === undefined && optCssClass === undefined) {
+    if (typeof optJsClass === 'undefined' &&
```
Comment (samthor) [inline src/mdlComponentHandler.js]: I've no strong opinions on this vs the `=== undefined` check, although I haven't read the style guide on this bit lately.

**Comment by samthor** [comment]
LGTM modulo nits. Do you want to get #1547 in too?

**Comment by sgomes** [comment]
Thanks for the review, @samthor! I'll add those changes in, squash and commit.

I'll review #1547 for you after this, but it's probably best to keep them separate for now. I suspect we'll need yet another PR to fix some things up after they're both merged (we're probably missing some exports/externs).

File: src/mdlComponentHandler.js (hunk 5 above)
Comment (sgomes) [inline src/mdlComponentHandler.js]: I think it's probably unnecessary at the moment, but I'll keep it for the sake of bullet-proofing :)

File: utils/jscs-rules/closure-camel-case.js (hunk 4 above)
Comment (sgomes) [inline utils/jscs-rules/closure-camel-case.js]: Fixed.

File: src/third_party/rAF.js (hunk 3 above)
Comment (sgomes) [inline src/third_party/rAF.js]: Fixed.

File: src/layout/layout.js (hunk 2 above)
Comment (sgomes) [inline src/layout/layout.js]: Fixed both.

File: gulpfile.js (hunk 1 above)
Comment (sgomes) [inline gulpfile.js]: Silly docblockr. Thanks!

**Pull request merged**